### DISCUSSION TITLE
Semantic versioning & breaking changes

### DISCUSSION THREAD

**Discussion by helmesjo** [stephenberry/glaze#1072] [General]
Hi!
I'm maintaining the [`build2` package of `glaze`](https://github.com/build2-packaging/glaze), and I noticed that in the bump to [`v2.8.0`](https://github.com/stephenberry/glaze/releases/tag/v2.8.0) there are breaking changes (I also tested in another project depending on the `build2` package of `glaze`, and indeed it broke the build).

Given that proper semantic versioning "demands" that only major version bumps allows breaking changes, and consequently version constraint conventions in different package managers rely on this assumption, I think it would be great if you got reserve breaking changes only for major version bumps. :)

Thanks in advance!

**Comment by stephenberry** [comment]
Given that Glaze is under rapid development with a broad API, breaking changes to some corner of the library are more common than most projects. I would be releasing a new major version every other week (or sometimes every other day) if I followed this rule and it would make it confusing for versioning more major changes. Hence, I try to reserve breaking changes to minor versions and explicitly state in the releases when breaking changes happen.

I do think breaking changes should be kept to a minimum, and typically the breaking changes only affect a small subset of users (for example most breaking changes right now are occurring in the REPE implementation, which isn't used by most users). In the case of [v2.8.0](https://github.com/stephenberry/glaze/releases/tag/v2.8.0) I do think this could have been a bump to 3.0.0 because it affects most users, so I'll keep your request in mind and try to release major versions for these kinds of changes.

Thanks for the feedback!

**Reply by helmesjo** [reply]
Sounds good, thanks!

**Reply by helmesjo** [reply]
(just some context that I forgot to emphasize, which was the main point):

Usually you have a a constraint like `glaze ^2.6.9` (meaning `<3.0.0, >=2.6.9`). Now if there are breaking changes in say `2.9.0`, this means that all old (and immutable) versions of the depender now are (potentially) broken (which can mean that already released software can't be built from source at it's release tag).

So it's not so much about the ability to modify code to work with the new version(s), but more about the _inability_ to modify old code.

Anyhow, no need to reply just wanted to add the scenario I had.

Thanks again!
